compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -155,26 +156,85 @@ impl UnderspecifiedArgKind {
     }
 }
 
-struct ClosureEraser<'tcx> {
-    tcx: TyCtxt<'tcx>,
+struct ClosureEraser<'a, 'tcx> {
+    infcx: &'a InferCtxt<'tcx>,
+    // When recursing into types, if an ADT has type parameters with a default type we do *not*
+    // want to replace that type parameter with `_`, as it will cause the normally hidden type
+    // parameter to be rendered. The best example of this is `Vec<T, Alloc>`, which we want to
+    // render as `Vec<T>` and not `Vec<T, _>` when `T` is unknown.
+    do_not_hide_nested_type: bool,
 }
 
-impl<'tcx> TypeFolder<TyCtxt<'tcx>> for ClosureEraser<'tcx> {
+impl<'a, 'tcx> ClosureEraser<'a, 'tcx> {
+    fn new_infer(&mut self) -> Ty<'tcx> {
+        self.infcx.next_ty_var(DUMMY_SP)
+    }
+}
+
+impl<'a, 'tcx> TypeFolder<TyCtxt<'tcx>> for ClosureEraser<'a, 'tcx> {
     fn cx(&self) -> TyCtxt<'tcx> {
-        self.tcx
+        self.infcx.tcx
     }
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
-        match ty.kind() {
+        let prev = self.do_not_hide_nested_type;
+        let ty = match ty.kind() {
             ty::Closure(_, args) => {
+                // For a closure type, we turn it into a function pointer so that it gets rendered
+                // as `fn(args) -> Ret`.
                 let closure_sig = args.as_closure().sig();
                 Ty::new_fn_ptr(
-                    self.tcx,
-                    self.tcx.signature_unclosure(closure_sig, hir::Safety::Safe),
+                    self.cx(),
+                    self.cx().signature_unclosure(closure_sig, hir::Safety::Safe),
                 )
             }
-            _ => ty.super_fold_with(self),
-        }
+            ty::Adt(def, _) => {
+                let generics = self.cx().generics_of(def.did());
+                if generics.own_params.iter().any(|param| param.default_value(self.cx()).is_some())
+                {
+                    // We have a type that has default types, like the allocator in Vec. We decided
+                    // to show `Vec` itself, because it hasn't yet been replaced by an `_` `Infer`,
+                    // but we want to ensure that the type parameter with default types does *not*
+                    // get replaced with `_` because then we'd end up with `Vec<_, _>`, instead of
+                    // `Vec<_>`.
+                    self.do_not_hide_nested_type = true;
+                    ty.super_fold_with(self)
```

Comment:
If there's a way to `fold` each param individually at this stage, then we could make it so we *only* avoid the type params like `Vec<_, Alloc>` and nothing else and we can likely get rid of the `bool` flag on `self` (if we can also make it so that `const`s don't get type erased too).

Reply:
if i understand u correctly, u can zip the generics with the `ty::Adt(_, args)` for that :thinking: you can then fold each arg manually and use `Ty::new_adt` instead of relying on `super_fold_with`

Reply:
Let me know if this is the right approach for building that param list.